File: crisp.py
#!/usr/bin/env python3
"""
crisp - generate SVG link text that bypasses GitHub's README underline.

Why this exists:
    GitHub's markdown CSS forces an underline on any <a> that contains text.
    But links that contain only <img> elements get no underline. So we render
    the project name as an SVG image and drop it inside an <a> tag. Clean,
    crisp link text. No CSS hacks. No sanitizer fights.

Usage:
    crisp.py "Retina" --color D97757 --link https://example.com
    crisp.py --batch projects.json
    crisp.py "Retina" --color D97757 --font /path/to/Inter-Bold.ttf

The width of the SVG is measured against Helvetica Bold (or your --font choice)
using Pillow. GitHub renders the SVG against the viewer's font stack at display
time. For short names this matches almost exactly; for long names you may want
to bump --trailing by a px or two.
"""

import argparse
import html
import json
import math
import os
import re
import sys
from pathlib import Path

try:
    from PIL import ImageFont
except ImportError:
    sys.stderr.write(
        "error: Pillow is required.\n"
        "  pip install Pillow\n"
    )
    sys.exit(2)

# fontTools is only needed for gradient rendering (it converts glyphs to vector
# paths so the gradient applies cleanly without the per-glyph fuzziness that
# browsers produce for <text fill="url(#gradient)">).
try:
    from fontTools.ttLib import TTFont
    from fontTools.pens.svgPathPen import SVGPathPen
    from fontTools.pens.transformPen import TransformPen
except ImportError:
    TTFont = None


SVG_TEMPLATE = (
    '<svg xmlns="http://www.w3.org/2000/svg" '
    'width="{w}" height="{h}" viewBox="0 0 {w} {h}">'
    '<text x="{x}" y="{y}" '
    'font-family="-apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif" '
    'font-size="{fs}" font-weight="{fw}" fill="{color}">{name}</text></svg>\n'
)

SVG_TEMPLATE_PATHS = (
    '<svg xmlns="http://www.w3.org/2000/svg" '
    'width="{w}" height="{h}" viewBox="0 0 {w} {h}">'
    '<defs>{grad}</defs>'
    '<g fill="url(#crisp-grad)" '
    'transform="translate({x}, {y}) scale({scale:.6f}, -{scale:.6f})">'
    '<path d="{d}"/></g></svg>\n'
)

GRADIENT_PRESETS = {
    "rainbow": ["#FF6B6B", "#FFA500", "#FFD700", "#10B981", "#4F46E5", "#A855F7"],
    "sunset": ["#DC2626", "#F59E0B", "#EC4899"],
    "ocean": ["#06B6D4", "#3B82F6", "#8B5CF6"],
    "mint": ["#10B981", "#06B6D4"],
    "candy": ["#EC4899", "#8B5CF6"],
    "dusk": ["#4F46E5", "#A855F7"],
}


FONT_CANDIDATES = [
    ("/System/Library/Fonts/Helvetica.ttc", 1),
    ("/System/Library/Fonts/HelveticaNeue.ttc", 1),
    ("/Library/Fonts/Helvetica.ttc", 1),
    ("/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf", 0),
    ("/usr/share/fonts/dejavu/DejaVuSans-Bold.ttf", 0),
    ("C:/Windows/Fonts/arialbd.ttf", 0),
]


def load_font(font_path, font_size):
    if font_path:
        return ImageFont.truetype(font_path, font_size)
    for path, index in FONT_CANDIDATES:
        if os.path.exists(path):
            try:
                return ImageFont.truetype(path, font_size, index=index)
            except Exception:
                continue
    sys.stderr.write(
        "warning: no bold system font found; falling back to Pillow default. "
        "Width measurement will be inaccurate. Pass --font /path/to/Bold.ttf for accuracy.\n"
    )
    return ImageFont.load_default()


def find_path_font(user_font_path, user_font_index):
    """Return (path, index) of a font file usable by fontTools for path-mode."""
    if user_font_path:
        return user_font_path, user_font_index or 0
    for path, index in FONT_CANDIDATES:
        if os.path.exists(path):
            return path, index
    return None, 0


def slugify(name):
    s = re.sub(r"[^a-z0-9]+", "-", name.lower())
    return s.strip("-") or "name"


def normalize_color(c):
    c = c.strip().lstrip("#")
    if not re.fullmatch(r"[0-9a-fA-F]{3}|[0-9a-fA-F]{6}", c):
        raise ValueError(f"invalid hex color: {c!r}")
    return "#" + c.upper()


def parse_gradient(spec):
    if isinstance(spec, list):
        return [normalize_color(c) for c in spec]
    spec = spec.strip()
    if spec.lower() in GRADIENT_PRESETS:
        return [normalize_color(c) for c in GRADIENT_PRESETS[spec.lower()]]
    return [normalize_color(c) for c in spec.split(",") if c.strip()]


def _build_path_gradient_def(colors, x_max):
    """userSpaceOnUse gradient spanning x=0..x_max horizontally (left → right).
    Coordinates are in the inner font-unit space established by the <g> transform."""
    n = len(colors)
    stops = "".join(
        f'<stop offset="{0 if n == 1 else int(round(i * 100 / (n - 1)))}%" stop-color="{c}"/>'
        for i, c in enumerate(colors)
    )
    return (
        f'<linearGradient id="crisp-grad" gradientUnits="userSpaceOnUse" '
        f'x1="0" y1="0" x2="{x_max}" y2="0">{stops}</linearGradient>'
    )


def _generate_paths(name, font_path, font_index, font_size,
                    height, leading, trailing, colors):
    """Convert glyphs to vector paths via fontTools, then fill with the gradient.

    Browsers render text-with-gradient via a fallback path that loses font
    hinting and looks fuzzy at small sizes. Pre-baking glyphs as paths and
    filling them via userSpaceOnUse gradient produces sharp output that
    matches the rendering of any other vector shape on the page.
    """
    if TTFont is None:
        raise RuntimeError(
            "fontTools is required for gradient rendering. "
            "Install it: pip install fonttools"
        )
    font_path_resolved, font_idx = find_path_font(font_path, font_index)
    if font_path_resolved is None:
        raise RuntimeError(
            "no system font found for path-mode rendering; "
            "pass --font /path/to/Bold.ttf"
        )

    tt = TTFont(font_path_resolved, fontNumber=font_idx)
    cmap = tt.getBestCmap()
    glyph_set = tt.getGlyphSet()
    hmtx = tt['hmtx']
    units_per_em = tt['head'].unitsPerEm
    scale = font_size / units_per_em

    # Accumulate glyph outlines into a single <path> with x-offsets baked into
    # the coordinates. One path means one bounding box, so the userSpaceOnUse
    # gradient spans the whole word instead of cycling per glyph.
    pen = SVGPathPen(glyph_set)
    x_pen = 0
    for char in name:
        glyph_name = cmap.get(ord(char)) or '.notdef'
        glyph = glyph_set[glyph_name]
        translated_pen = TransformPen(pen, (1, 0, 0, 1, x_pen, 0))
        glyph.draw(translated_pen)
        x_pen += hmtx[glyph_name][0]
    combined_d = pen.getCommands()

    width = int(round(leading + x_pen * scale + trailing))
    # See generate() for baseline rationale (align="absmiddle" alignment).
    baseline = round(height / 2 + font_size * 0.25)
    grad_def = _build_path_gradient_def(colors, x_pen)

    svg = SVG_TEMPLATE_PATHS.format(
        w=width, h=height, x=leading, y=baseline, scale=scale,
        grad=grad_def, d=combined_d,
    )
    return svg, width


def generate(name, color, gradient, gradient_angle, font_path, font_index,
             font_size, font_weight, height, leading, trailing):
    if gradient:
        # Gradient → path-based rendering (sharp glyphs, clean color spread)
        colors = parse_gradient(gradient)
        return _generate_paths(
            name, font_path, font_index, font_size,
            height, leading, trailing, colors,
        )

    # Solid color → keep <text>. Browsers use the native font rasterizer
    # (with hinting) for single-color text, which is sharper than any path
    # conversion at small sizes.
    font = load_font(font_path, font_size)
    text_width = font.getlength(name)
    width = int(round(text_width + leading + trailing))
    # Baseline tuned for align="absmiddle" inline rendering:
    # vertical-align:middle anchors the image center to body x-height middle
    # (~font_size * 0.25 above body baseline). Putting the SVG baseline at
    # height/2 + font_size * 0.25 makes the SVG text baseline land exactly on
    # body baseline, so the text sits flush with surrounding paragraph.
    baseline = round(height / 2 + font_size * 0.25)
    svg = SVG_TEMPLATE.format(
        w=width, h=height, x=leading, y=baseline,
        fs=font_size, fw=font_weight, color=color,
        name=html.escape(name, quote=True),
    )
    return svg, width


def render_snippet(name, output_path, link):
    img_tag = (
        f'<img src="{html.escape(output_path, quote=True)}" '
        f'align="absmiddle" alt="{html.escape(name, quote=True)}" />'
    )
    if link:
        return f'<a href="{html.escape(link, quote=True)}">{img_tag}</a>'
    return img_tag


def process_one(args, item):
    name = item["name"]
    gradient = item.get("gradient") or args.gradient
    gradient_angle = item.get("gradient_angle") if "gradient_angle" in item else args.gradient_angle
    color = None if gradient else normalize_color(item.get("color", args.color))
    output = item.get("output") or args.output or f"{slugify(name)}.svg"
    link = item.get("link") or args.link
    font_path = item.get("font") or args.font
    font_index = item.get("font_index") if "font_index" in item else args.font_index
    font_size = item.get("font_size") or args.font_size
    font_weight = item.get("font_weight") or args.font_weight
    height = item.get("height") or args.height
    leading = item.get("leading") if "leading" in item else args.leading
    trailing = item.get("trailing") if "trailing" in item else args.trailing

    svg, width = generate(
        name=name,
        color=color,
        gradient=gradient,
        gradient_angle=gradient_angle,
        font_path=font_path,
        font_index=font_index,
        font_size=font_size,
        font_weight=font_weight,
        height=height,
        leading=leading,
        trailing=trailing,
    )
    out_path = Path(output)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_text(svg, encoding="utf-8")
    snippet = render_snippet(name, str(out_path), link)
    return out_path, width, snippet


def build_parser():
    p = argparse.ArgumentParser(
        prog="crisp",
        description="Generate SVG link text that bypasses GitHub's README underline.",
    )
    p.add_argument("name", nargs="?", help="Display text (e.g., the project name).")
    p.add_argument("--color", default="0969DA", help="Hex color of the text. Default: 0969DA (GitHub link blue). Ignored if --gradient is set.")
    p.add_argument("--gradient", help="Linear gradient as comma-separated hex colors (e.g., 'EC4899,8B5CF6') or preset name (rainbow, sunset, ocean, mint, candy, dusk).")
    p.add_argument("--gradient-angle", type=int, default=90, help="Gradient angle in degrees, CSS-style. 0=up, 90=right (horizontal, default), 180=down, 270=left.")
    p.add_argument("--output", help="Output SVG path. Default: <slugified-name>.svg in cwd.")
    p.add_argument("--link", help="Wrap the snippet in <a href=...>. Optional.")
    p.add_argument("--font", help="Path to a TTF/OTF/TTC font file. Default: auto-detect Helvetica Bold.")
    p.add_argument("--font-index", type=int, default=None, help="Font index inside a TTC collection. Default: 0 if user-supplied, 1 for system Helvetica.ttc (Bold).")
    p.add_argument("--font-size", type=int, default=16, help="Font size in px. Default: 16.")
    p.add_argument("--font-weight", type=int, default=600, help="Font weight. Default: 600.")
    p.add_argument("--height", type=int, default=22, help="SVG height in px. Default: 22.")
    p.add_argument("--leading", type=int, default=6, help="Leading padding in px. Default: 6.")
    p.add_argument("--trailing", type=int, default=4, help="Trailing padding in px. Default: 4.")
    p.add_argument("--batch", help="Path to a JSON array of items. Each item may set any of: name, color, output, link, font, font_size, font_weight, height, leading, trailing.")
    p.add_argument("--quiet", action="store_true", help="Suppress the stdout snippet (only write files).")
    return p


def main(argv=None):
    args = build_parser().parse_args(argv)

    if args.batch:
        with open(args.batch, encoding="utf-8") as f:
            items = json.load(f)
        if not isinstance(items, list):
            sys.stderr.write("error: --batch file must contain a JSON array.\n")
            sys.exit(2)
    elif args.name:
        items = [{"name": args.name}]
    else:
        build_parser().print_help(sys.stderr)
        sys.exit(2)

    for item in items:
        if "name" not in item or not item["name"]:
            sys.stderr.write(f"error: item missing 'name': {item!r}\n")
            sys.exit(2)
        try:
            out_path, width, snippet = process_one(args, item)
        except ValueError as e:
            sys.stderr.write(f"error: {e}\n")
            sys.exit(2)
        if not args.quiet:
            print(snippet)
        sys.stderr.write(f"wrote {out_path} ({width}x{args.height})\n")


if __name__ == "__main__":
    main()

File: docs/crisp.js
/**
 * crisp.js — browser port of the crisp-links Python CLI.
 *
 * Generates the same SVGs as `crisp.py`. Width is measured via Canvas
 * with the same system font stack the SVGs render against, so widths
 * stay within 1-2px of the Python (Pillow + Helvetica Bold) version —
 * comfortably inside the leading/trailing buffer.
 *
 * No DOM dependencies. Safe to require/import anywhere a Canvas is available.
 */
(function (global) {
  "use strict";

  const PRESETS = {
    rainbow: ["#FF6B6B", "#FFA500", "#FFD700", "#10B981", "#4F46E5", "#A855F7"],
    sunset:  ["#DC2626", "#F59E0B", "#EC4899"],
    ocean:   ["#06B6D4", "#3B82F6", "#8B5CF6"],
    mint:    ["#10B981", "#06B6D4"],
    candy:   ["#EC4899", "#8B5CF6"],
    dusk:    ["#4F46E5", "#A855F7"],
  };

  const DEFAULTS = Object.freeze({
    fontSize: 16,
    fontWeight: 600,
    height: 22,
    leading: 6,
    trailing: 4,
    gradientAngle: 90,
    fontFamily: '-apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif',
  });

  // Reusable canvas for text measurement
  let _ctx = null;
  function _measureCtx() {
    if (_ctx) return _ctx;
    const canvas = (typeof OffscreenCanvas !== "undefined")
      ? new OffscreenCanvas(8, 8)
      : (typeof document !== "undefined" ? document.createElement("canvas") : null);
    if (!canvas) throw new Error("crisp.js needs a Canvas (browser or OffscreenCanvas).");
    _ctx = canvas.getContext("2d");
    return _ctx;
  }

  // ---- Path-mode font (opentype.js) -------------------------------------
  // Used for gradient rendering: converts glyphs to vector paths so the
  // gradient applies cleanly without per-glyph fuzziness.
  let _font = null;
  let _fontLoadPromise = null;

  function loadFont(url) {
    if (_font) return Promise.resolve(_font);
    if (_fontLoadPromise) return _fontLoadPromise;
    if (typeof opentype === "undefined") {
      return Promise.reject(new Error("opentype.js not loaded"));
    }
    _fontLoadPromise = fetch(url)
      .then((r) => {
        if (!r.ok) throw new Error("font fetch failed: " + r.status);
        return r.arrayBuffer();
      })
      .then((buf) => {
        _font = opentype.parse(buf);
        return _font;
      })
      .catch((err) => {
        _fontLoadPromise = null; // allow retry
        throw err;
      });
    return _fontLoadPromise;
  }

  function isFontLoaded() {
    return _font !== null;
  }

  function measureWidth(text, fontSize, fontWeight, fontFamily) {
    const ctx = _measureCtx();
    ctx.font = `${fontWeight} ${fontSize}px ${fontFamily}`;
    return ctx.measureText(text).width;
  }

  function normalizeColor(input) {
    if (input == null) throw new Error("missing color");
    let c = String(input).trim().replace(/^#/, "");
    if (!/^([0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/.test(c)) {
      throw new Error(`invalid hex color: ${input}`);
    }
    if (c.length === 3) c = c.split("").map((ch) => ch + ch).join("");
    return "#" + c.toUpperCase();
  }

  function parseGradient(spec) {
    if (Array.isArray(spec)) {
      return spec.map(normalizeColor);
    }
    if (typeof spec !== "string") throw new Error("gradient must be string or array");
    const trimmed = spec.trim();
    if (PRESETS[trimmed.toLowerCase()]) {
      return PRESETS[trimmed.toLowerCase()].map(normalizeColor);
    }
    return trimmed.split(",").map((s) => s.trim()).filter(Boolean).map(normalizeColor);
  }

  // CSS-style angle: 0=up, 90=right, 180=down, 270=left.
  // Returns endpoints in 0-100 percentages.
  function gradientEndpoints(angleDeg) {
    const rad = (angleDeg * Math.PI) / 180;
    const dx = Math.sin(rad);
    const dy = -Math.cos(rad);
    return {
      x1: 50 - dx * 50,
      y1: 50 - dy * 50,
      x2: 50 + dx * 50,
      y2: 50 + dy * 50,
    };
  }

  function buildGradientDef(colors, angle) {
    const { x1, y1, x2, y2 } = gradientEndpoints(angle);
    const n = colors.length;
    const stops = colors
      .map((c, i) => {
        const offset = n === 1 ? 0 : Math.round((i * 100) / (n - 1));
        return `<stop offset="${offset}%" stop-color="${c}"/>`;
      })
      .join("");
    return (
      `<linearGradient id="crisp-grad" x1="${x1.toFixed(1)}%" y1="${y1.toFixed(1)}%" ` +
      `x2="${x2.toFixed(1)}%" y2="${y2.toFixed(1)}%">${stops}</linearGradient>`
    );
  }

  function xmlEscape(s) {
    return String(s)
      .replace(/&/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;")
      .replace(/"/g, "&quot;")
      .replace(/'/g, "&apos;");
  }

  // Path-based generator: converts text to vector paths via opentype.js, then
  // fills with userSpaceOnUse gradient. Avoids the fuzzy rendering browsers
  // produce for <text fill="url(#g)"> at small sizes. Requires loadFont() first.
  function generatePaths(opts) {
    if (!_font) throw new Error("font not loaded; call Crisp.loadFont(url) first");

    const name = opts.name;
    const fontSize = opts.fontSize ?? DEFAULTS.fontSize;
    const height = opts.height ?? DEFAULTS.height;
    const leading = opts.leading ?? DEFAULTS.leading;
    const trailing = opts.trailing ?? DEFAULTS.trailing;
    const colors = parseGradient(opts.gradient);
    const angle = opts.gradientAngle ?? DEFAULTS.gradientAngle;

    const baseline = Math.round(height / 2 + fontSize * 0.25);
    const advance = _font.getAdvanceWidth(name, fontSize);
    const width = Math.round(leading + advance + trailing);

    // opentype.js positions glyphs at (x, y) with y = baseline in display coords.
    // The path data comes out already positioned and scaled — no extra transforms.
    const path = _font.getPath(name, leading, baseline, fontSize);
    const pathData = path.toPathData(3); // 3 decimal precision

    const n = colors.length;
    const stops = colors
      .map((c, i) => {
        const offset = n === 1 ? 0 : Math.round((i * 100) / (n - 1));
        return `<stop offset="${offset}%" stop-color="${c}"/>`;
      })
      .join("");

    // userSpaceOnUse so the gradient spans the entire word, not per-glyph.
    // Default direction is horizontal (leading → leading+advance). For non-90
    // angles, apply a gradientTransform rotation around the gradient center.
    const gradX1 = leading;
    const gradX2 = leading + advance;
    const cx = (gradX1 + gradX2) / 2;
    const rotation = angle - 90;
    const gradientTransform =
      rotation === 0
        ? ""
        : ` gradientTransform="rotate(${rotation} ${cx} ${baseline})"`;

    const gradientDef =
      `<linearGradient id="crisp-grad" gradientUnits="userSpaceOnUse" ` +
      `x1="${gradX1}" y1="0" x2="${gradX2}" y2="0"${gradientTransform}>` +
      stops +
      `</linearGradient>`;

    const svg =
      `<svg xmlns="http://www.w3.org/2000/svg" ` +
      `width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">` +
      `<defs>${gradientDef}</defs>` +
      `<path d="${pathData}" fill="url(#crisp-grad)"/>` +
      `</svg>`;

    return { svg, width, height };
  }

  // Solid-color path: keeps <text> so the browser uses its native font
  // rasterizer (with hinting). Smaller, sharper than path conversion at 16px.
  function generateText(opts) {
    const name = opts.name;
    const fontSize = opts.fontSize ?? DEFAULTS.fontSize;
    const fontWeight = opts.fontWeight ?? DEFAULTS.fontWeight;
    const height = opts.height ?? DEFAULTS.height;
    const leading = opts.leading ?? DEFAULTS.leading;
    const trailing = opts.trailing ?? DEFAULTS.trailing;
    const fontFamily = opts.fontFamily ?? DEFAULTS.fontFamily;

    const textWidth = measureWidth(name, fontSize, fontWeight, fontFamily);
    const width = Math.round(textWidth + leading + trailing);
    // See above for align="absmiddle" baseline rationale.
    const baseline = Math.round(height / 2 + fontSize * 0.25);

    let gradientDef = "";
    let fill;
    if (opts.gradient) {
      // Fallback path (no opentype font loaded yet). Uses <text> + gradient
      // which is fuzzier but renders without the font dependency.
      const colors = parseGradient(opts.gradient);
      const angle = opts.gradientAngle ?? DEFAULTS.gradientAngle;
      gradientDef = `<defs>${buildGradientDef(colors, angle)}</defs>`;
      fill = "url(#crisp-grad)";
    } else {
      const color = normalizeColor(opts.color ?? "#0969DA");
      fill = color;
    }

    const escapedName = xmlEscape(name);
    const svg =
      `<svg xmlns="http://www.w3.org/2000/svg" ` +
      `width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">` +
      gradientDef +
      `<text x="${leading}" y="${baseline}" ` +
      `font-family="${xmlEscape(fontFamily)}" ` +
      `font-size="${fontSize}" font-weight="${fontWeight}" fill="${fill}">` +
      `${escapedName}</text></svg>`;

    return { svg, width, height };
  }

  function generate(opts) {
    if (!opts || !opts.name) throw new Error("name is required");
    // Gradient + font loaded → path-based (sharp). Otherwise text-based.
    if (opts.gradient && _font) {
      return generatePaths(opts);
    }
    return generateText(opts);
  }

  function slugify(name) {
    const s = String(name).toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");
    return s || "name";
  }

  function snippet({ filename, name, link }) {
    const file = xmlEscape(filename);
    const alt = xmlEscape(name);
    const img = `<img src="${file}" align="absmiddle" alt="${alt}" />`;
    if (link) {
      return `<a href="${xmlEscape(link)}">${img}</a>`;
    }
    return img;
  }

  function downloadSvg(svgString, filename) {
    const blob = new Blob([svgString], { type: "image/svg+xml;charset=utf-8" });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    setTimeout(() => URL.revokeObjectURL(url), 250);
  }

  function svgDataUrl(svgString) {
    return "data:image/svg+xml;utf8," + encodeURIComponent(svgString);
  }

  global.Crisp = {
    PRESETS,
    DEFAULTS,
    measureWidth,
    normalizeColor,
    parseGradient,
    gradientEndpoints,
    buildGradientDef,
    generate,
    generateText,
    generatePaths,
    loadFont,
    isFontLoaded,
    slugify,
    snippet,
    downloadSvg,
    svgDataUrl,
  };
})(typeof window !== "undefined" ? window : globalThis);
